apps/condo/domains/common/components/Prompt.tsx

```diff
@@ -21,51 +21,99 @@ const Prompt: React.FC<IPromptProps> = ({ children, title, form, handleSave: for
     const [isModalVisible, setIsModalVisible] = useState(false)
     const isIgnoringPrompt = useRef(false)
     const initialFormState = useRef({})
+    const isBackButtonClicked = useRef(false)
     const router = useRouter()
+
     const showModal = () => setIsModalVisible(true)
     const hideModal = () => setIsModalVisible(false)
-    const handleCancel = () => {
+    const handleCancel = async () => {
         isIgnoringPrompt.current = true
         hideModal()
-        router.push(next)
+
+        if (isBackButtonClicked.current) {
+            router.back()
+        }
+
+        await router.push(next)
     }
-    const handleSave = () => {
-        isIgnoringPrompt.current = true
+    const handleSave = async () => {
         hideModal()
-        formSubmit()
+        const isValid = await form.validateFields()
+
+        if (isValid) {
+            isIgnoringPrompt.current = true
+            formSubmit()
+
+            if (isBackButtonClicked.current) {
+                router.back()
+            }
+        }
     }
     // TODO(zuch): find a solution to watch for file changes and contact changes
     // as they are using custom hooks and their fields do not present on initial form state
     const isFormChanged = () => {
         const newFormFields = pick(form.getFieldsValue(), Object.keys(initialFormState.current))
         return !isEqual(initialFormState.current, newFormFields)
     }
+
     useEffect(() => {
         initialFormState.current = form.getFieldsValue()
+    }, [])
+
+    useEffect(() => {
```

Comment:
Why did you separate these `useEffect`'s?

Reply:
At some point there were different dependencies, returned to one